src/stream.rs

```diff
@@ -0,0 +1,533 @@
+    pub fn len(&self) -> u64 {
+        self.length
+    }
+
+    /// Append data to the end of the stream.
+    ///
+    /// Returns the new stream length.
+    ///
+    pub fn append(&mut self, mut data: &[u8]) -> Result<u64, StreamError> {
+        let length = data.len();
+        let buffered = self.length as usize & Self::block_mask();
+        if buffered + length <= Self::block_size() {
+            // All the data will fit in the existing buffer.
+            unsafe { self.write_to_buffer(data, length) };
+            if self.length as usize & Self::block_mask() == 0 {
+                // Buffer is now full, and can be written to file.
+                unsafe { self.write_buffer_to_file()? };
+            }
+        } else {
+            // The data will fill the existing buffer.
+            if buffered > 0 {
+                // The buffer is partly used. Fill it and write it out.
+                let length = Self::block_size() - buffered;
+                unsafe { self.write_to_buffer(data, length) };
+                // Buffer is now full, and can be written to file.
+                unsafe { self.write_buffer_to_file()? };
+                data = &data[length..];
+            }
+            // The buffer is curently empty, so we are free to write whole
+            // blocks of data directly to the file, bypassing the buffer.
+            let direct = data.len() & !Self::block_mask();
+            if direct > 0 {
+                unsafe { self.write_to_file(&data[..direct])? };
+                data = &data[direct..];
+                self.length += direct as u64;
+            }
+            let length = data.len();
+            if length > 0 {
+                // There is data remaining which will fit in the next buffer.
+                unsafe { self.write_to_buffer(data, length) };
+            }
+        }
+        // Update shared length value, and return new length.
+        self.shared.length.store(self.length, Release);
+        Ok(self.length)
+    }
+
+    /// Helper method for writing data to buffer.
+    ///
+    /// The data must be guaranteed to fit within the remaining space.
+    ///
+    #[inline(always)]
+    unsafe fn write_to_buffer(&mut self, data: &[u8], length: usize) {
+        copy_nonoverlapping(data.as_ptr(), self.ptr, length);
+        self.ptr = self.ptr.add(length);
+        self.length += length as u64;
+    }
+
+    /// Helper method for writing buffer to file.
+    ///
+    /// The buffer must be guaranteed to be full.
+    #[inline(always)]
+    unsafe fn write_buffer_to_file(&mut self) -> Result<(), StreamError> {
+        let buf = slice::from_raw_parts(self.buf, Self::block_size());
+        self.write_to_file(buf)
+    }
+
+    /// Helper method for writing data to file.
+    ///
+    /// The data must be guaranteed to be a multiple of the block size.
+    ///
+    unsafe fn write_to_file(&mut self, data: &[u8]) -> Result<(), StreamError> {
```

Comment:
This function is a little annoying: it is `unsafe` to call because it requires a guarantee that the data length is a multiple of the block size, but the code within it is actually entirely safe.

I understand that Rust plans to eventually have `unsafe` functions not automatically mark their scope `unsafe`, but it's not done yet: https://github.com/rust-lang/rust/issues/71668

Reply:
Oh, actually we can have that now with `#[deny(unsafe_op_in_unsafe_fn)]`.